pkg/controller/cluster/controller.go

```diff
@@ -473,3 +516,52 @@ func (e *external) getPayload(ctx context.Context, ref *v1alpha1.SecretReference
 
 	return nil, nil
 }
+
+// ExtractKubeconfigFromSecretRef extracts login information from a Kubeconfig Secret Reference
+func (e *external) ExtractKubeconfigFromSecretRef(ctx context.Context, p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Cluster) error {
+
+	kubeconfig, err := e.getPayload(ctx, p.Config.KubeconfigSecretRef)
+	if err != nil {
+		return err
+	}
+	restConfig, err := NewRESTConfigForKubeConfig(kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, errParseKubeconfig)
+	}
+
+	if restConfig.Host != "" {
+		if p.Name == "" {
+			p.Name = restConfig.Host
+			r.Name = restConfig.Host
+		}
+		p.Server = restConfig.Host
+		r.Server = restConfig.Host
+	}
+
+	if restConfig.BearerToken != "" {
+		r.Config.BearerToken = restConfig.BearerToken
+	}
+
+	if restConfig.Username != "" {
+		p.Config.Username = &restConfig.Username
+		r.Config.Username = restConfig.Username
+	}
+
+	r.Config.TLSClientConfig = argocdv1alpha1.TLSClientConfig{
+		Insecure:   restConfig.TLSClientConfig.Insecure,
+		CAData:     restConfig.CAData,
+		CertData:   restConfig.TLSClientConfig.CertData,
+		KeyData:    restConfig.TLSClientConfig.KeyData,
+		ServerName: restConfig.TLSClientConfig.ServerName,
+	}
+	return nil
+}
+
+// NewRESTConfigForKubeConfig returns a REST Config for the given KubeConfigValue.
+func NewRESTConfigForKubeConfig(kubeConfig []byte) (*rest.Config, error) {
```

Comment:
Should be lower case as its only used internally.

```suggestion
func NewRESTConfigForKubeConfig(kubeConfig []byte) (*rest.Config, error) {
```

Reply:
renamed to lowercase